### PR TITLE
4-import-the-movies-csv-into-the-database

### PR DESCRIPTION
**Pull request by selvipeb**
Issue 4: https://github.com/selvipeb/smartflix/issues/4

Build a Bootstrap-styled homepage displaying all the movies in the [provided CSV file](https://www.dropbox.com/s/iqqbb64z8eogqxb/netflix_titles.zip?dl=0).

In this ticket, you'll import that data into your application by storing it in a [Postgres](https://www.postgresql.org/) database.
* Install [Postgres](https://postgresapp.com/).
* [Connect](https://guides.rubyonrails.org/configuring.html#configuring-a-postgresql-database) your Rails application to your Postgres database.
* Create the database using the Rails CLI.
* Generate an appropriately-named Rails model (and associated migration) using the [Rails Command Line](https://guides.rubyonrails.org/command_line.html) with the following properties:
** show_type, as a string that is never null.
** title, as a string that is never null.
** director, as a string.
** cast, as a text.
** country, as a string.
** date_added, as a date.
** release_year, as an integer that is never null.
** rating, as a string.
** duration, as a string.
** listed_in, as a text.
** description, as a text.
* Feel free to rename the properties to follow conventions.
* Migrate the database to update it with the new structure.
* Create a new Rake task for importing the CSV. Consider that you'll be [importing a lot of data](https://mattboldt.com/importing-massive-data-into-rails/)!
* Run the Rake task to import the CSV data into your database. Check that it worked using the [Rails Console](https://guides.rubyonrails.org/command_line.html#bin-rails-console).
* Rewrite your controller and view to use the database-backed models, instead of reading from the CSV.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -5,8 +5,8 @@ ruby '2.7.2'
 
 # Bundle edge Rails instead: gem 'rails', github: 'rails/rails', branch: 'main'
 gem 'rails', '~> 6.1.4', '>= 6.1.4.4'
-# Use sqlite3 as the database for Active Record
-gem 'sqlite3', '~> 1.4'
+# Use postgresql as the database for Active Record
```
Comment (sjmog) [inline Gemfile:8]: I really like that you continued the established convention of commenting what each gem is for.

File: Gemfile [hunk 2]
```diff
@@ -58,4 +58,6 @@ group :development, :test do
   gem 'pry-rails'
   # Add rubocop for better rails codes
   gem 'rubocop'
+  # Add factory bot to create data effectively
```
Comment (sjmog) [inline Gemfile]: We refer to this kind of test data as "fixtures".

File: app/models/film.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+class Film < ApplicationRecord
+  MOVIE = 'Movie'.freeze
+  TV_SERIES = 'TV Show'.freeze
```
Comment (sjmog) [inline app/models/film.rb]: Really nice practice to store these important fields as constants and freeze the string. I wonder if I wouldn't have added a `_TYPE` suffix to make it clear which column it refers to.

File: bin/setup [hunk 4]
```diff
@@ -26,7 +26,7 @@ FileUtils.chdir APP_ROOT do
   # end
 
   puts "\n== Preparing database =="
-  system! 'bin/rails db:prepare'
+  system! 'bin/rails db:drop db:prepare'
```
Comment (sjmog) [inline bin/setup:29]: I love that you updated the setup scripts too. That's great attention to detail

File: lib/tasks/import_films_from_csv.rake [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        'listed_in': params['listed_in'],
+        'description': params['description']
+      }
+    end
+    
+    def create_film(args)
+      Film.import(**args)
+    end
+
+  end
```
Comment (sjmog) [inline lib/tasks/import_films_from_csv.rake:28]: This is a nice way to do this. I wonder if you could have extracted the class to its own file, in order to unit-test it.

File: spec/tasks/import_films_from_csv_spec.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+  let(:file_content) { file_fixture('flix.csv').read }
+  let(:task) { Rake::Task["import_films_from_csv:run"] }
+
+  it 'triggers Film.import from csv' do
+    expect(Film).to receive(:import).with(
+      a_hash_including(title: 'Dick Johnson Is Dead')
+    )
+    
+    subject
```
Comment (sjmog) [inline spec/tasks/import_films_from_csv_spec.rb:19]: I think it's reasonable to use this London style of testing (i.e. "test X receives Y"). Since you are including the `rails_helper` (which spins up the whole app anyway), you might have considered the Chicago style – checking that it actually creates the object in the DB.

I wonder if this test might also include considerations for edge cases – e.g. where a required field is missing. However, it tests the "happy path" well.

Lastly, I wonder if this might have been better expressed as a unit test for the object you created. But that's really a matter of personal preference.

**Review by sjmog — APPROVED** [review]
There are some stylistic differences in our approaches but I think this is a great solution :)

**Comment by sjmog** [comment]
You may need to add me as a collaborator to this repo for my review to "count" for merging :)

**Comment by selvipeb** [comment]
> You may need to add me as a collaborator to this repo for my review to "count" for merging :)

Thanks for the review, Sam! I've invited you as a collaborator 😄

**Review by sjmog — APPROVED** [review]

File: Gemfile (hunk 1 above)
Comment (selvipeb) [reply Gemfile:8]: Thank you 🙌

File: app/models/film.rb (hunk 3 above)
Comment (selvipeb) [reply app/models/film.rb]: Will update this 👍

File: bin/setup (hunk 4 above)
Comment (selvipeb) [reply bin/setup:29]: Thanks 😊

File: lib/tasks/import_films_from_csv.rake (hunk 5 above)
Comment (selvipeb) [reply lib/tasks/import_films_from_csv.rake:28]: Thanks! I have added the unit test for it [here](https://github.com/selvipeb/smartflix/pull/11/commits/d7655077ccaf867e768220600547076167de69a3)

File: spec/tasks/import_films_from_csv_spec.rb (hunk 6 above)
Comment (selvipeb) [reply spec/tasks/import_films_from_csv_spec.rb:19]: Thanks! I have added the unit test for it and also the spec counting Film record [here](https://github.com/selvipeb/smartflix/pull/11/commits/d7655077ccaf867e768220600547076167de69a3)
